### PR TITLE
perf: don't always return basis-vector array

### PR DESCRIPTION
**Pull request by steven-murray**
## Description

This makes it possible to _not_ return the basis vector array, which is often not needed in simulation contexts, and takes a reasonably significant amount of memory.

It also computes the modified az/za arrays in the `map_coordinates` interpolator in-place to reduce memory (this is fine since these arrays are previously copied outside this function).

## Motivation and Context

In profiling `fftvis`, I found that the memory usage was higher than naively expected (this was multiprocessing over 128 threads, so everything was magnified). One of the unnecessary memory allocations was this basis-vector array.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change
- [x] Performance enhancement

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Performance checklist:
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/CHANGELOG.md).

### CODE REVIEW

**Review by bhazelton — COMMENTED** [review]
This is great, thanks. Just a few suggestions.

File: tests/uvbeam/test_uvbeam.py
Comment (steven-murray) [reply tests/uvbeam/test_uvbeam.py:1285]: Good catch.

File: CHANGELOG.md
Comment (bhazelton) [inline CHANGELOG.md:38]: This needs to be moved up to be in the Unreleased section.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (bhazelton) [inline src/pyuvdata/uvbeam/uvbeam.py:1723]: This is going to break things. We need a deprecation process because this keyword didn't exist in the past. I think you should default it to None and then catch that case, issue a deprecation warning and set it to True. The deprecation warning should say that it will default to False starting in version 3.3.

File: CHANGELOG.md
Comment (steven-murray) [reply CHANGELOG.md:38]: Ah yeah, things were released in between ;-)

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (steven-murray) [reply src/pyuvdata/uvbeam/uvbeam.py:1723]: Ah yeah. So at the moment, I'm doing this deprecation stuff only in the top-level function `interp()` because these functions are private and therefore need not be API-stable. Is that OK or do you want me to do the deprecation logic everywhere?

Comment (bhazelton) [reply src/pyuvdata/uvbeam/uvbeam.py:1723]: Oh, sorry, I didn't see that. I think that's fine.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (bhazelton) [inline src/pyuvdata/uvbeam/uvbeam.py:2089]: This isn't right. The default is still False for this method, will become True in v3.3

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (bhazelton) [inline src/pyuvdata/uvbeam/uvbeam.py:2124]: This should be a deprecation warning.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (steven-murray) [reply src/pyuvdata/uvbeam/uvbeam.py:2089]: Ah, yes.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (steven-murray) [reply src/pyuvdata/uvbeam/uvbeam.py:2124]: Good catch.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (steven-murray) [reply src/pyuvdata/uvbeam/uvbeam.py:2089]: So, I've left the default for now as True, and added the statement that in v3.3 it will be updated to be False.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (bhazelton) [inline src/pyuvdata/uvbeam/uvbeam.py:1808]: Minor nitpick: I'd find these divide lines easier to read if they were multiplied by the reciprocal. I have to think harder about what the two divides mean.

**Review by bhazelton — COMMENTED** [review]
I think the tests will pass now if you rebase. We just released another version, so be careful with the changelog.

File: src/pyuvdata/uvbeam/uvbeam.py
Comment (steven-murray) [reply src/pyuvdata/uvbeam/uvbeam.py:1808]: Np. Updated.

**Review by bhazelton — DISMISSED** [review]
Looks great, thanks! I think it can be merged once tests pass.

**Comment by bhazelton** [comment]
@steven-murray There's some missing code coverage, specifically for the new deprecation warning you added.

The pyuvsim tests are failing because of the new warning, which is ok, I can patch that up after we merge this.

**Review by bhazelton — APPROVED** [review]

**Comment by bhazelton** [comment]
Actually, the pyuvsim error is fixed now that we've merged the pyuvsim PR that uses the BeamInterface objects.

**Pull request merged**
